### ISSUE TITLE
WebAuthn support

### ISSUE THREAD

**Issue by hex-m** [privacyidea/FreeRADIUS#62]
Is it possible to support WebAuthn via RADIUS?

From my understanding this may be hard - similar to [push tokens](https://github.com/privacyidea/FreeRADIUS/issues/38) where @cornelinux mentioned:

> The problem is that the out of band auth with push can not work well with a protocol like RADIUS.

**Comment by EchedelleLR** [comment]
If that is the thing, how is that YubiKey is supported in FreeRADIUS using PAM as backend?

https://developers.yubico.com/yubico-pam/YubiKey_and_FreeRADIUS_via_PAM.html

Would not be possible to use FreeRADIUS PAM authentication and implementing privacyIDEA support through its PAM plugin to bring FreeRAIUS with WebAuthn support?

**Comment by hex-m** [comment]
From your link:

> Two-factor legacy Username + password + YubiKey OTP authentication for RADIUS server

YubiKey-Devices support other protocols than FIDO2/WebAuthn.

**Comment by EchedelleLR** [comment]
In my case, I am only interested in WebAuthn since I use SoloKeys.

Edit: okay now I see what you say.

But if PI already supports PAM with FIDO and FreeRADIUS supports using PAM module, would not be possible?

**Comment by nilsbehlen** [comment]
hi, our pam does currently not support fido2. however, it can easily be implemented, it is just a matter of having the time.
if freeradius could use the pam module, that would be great and maybe a reason to implement fido2 sooner.
Do you know of any pam module or combination that makes fido2 work with freeradius?

**Comment by mattv8** [comment]
I'm going to watch this thread, because WebAuthn via RADIUS would be a nice feature to have. Wish I had the time to help develop it, or I'd contribute.

My use case is to FIDO2/WebAuthn OpenVPN connections (although I'm not sure if anything other than PAP is supported, so that could be another isue...).

**Comment by plettich** [comment]
The Problem with FreeRADIUS and WebAuthn/FIDO2 is that the client needs to support it as well (i.e. talk to the authenticator hardware and send the appropriate data via the RADIUS protocol).
The Yubico document about FreeRADIUS and PAM (archived here: https://github.com/Yubico/yubico-pam/blob/master/doc/YubiKey_and_FreeRADIUS_1FA_via_PAM.adoc ) uses the Yubikey OTP functionality, not FIDO2.

**Comment by mattv8** [comment]
>The Problem with FreeRADIUS and WebAuthn/FIDO2 is that the client needs to support it as well (i.e. talk to the authenticator hardware and send the appropriate data via the RADIUS protocol).

Yes you are right, but wouldn’t that be confirmed when the user enrolls their device, whether or not the client supports it? 

Currently, PrivacyIDEA is designed to function alongside your primary authentication database, not as a proxy between the client and the authentication server. I'm sure this has been suggested before, but one approach could be like how [DuoAuthProxy](https://duo.com/docs/authproxy-reference#use-the-proxy-manager) does it.

You'd build the RADIUS protocol into PrivacyIDEA then use PrivacyIDEA as a proxy for the FIDO2 flow: PrivacyIDEA would implement a RADIUS server plugin that translates RADIUS Access‑Request attributes into WebAuthn registration and authentication challenges (using a Python library like python‑fido2), persist the challenge state in its own database, then proxy the corresponding RADIUS Access‑Challenge and Access‑Accept responses containing the client’s attestation or assertion data to the terminal auth server.

I realize building a full RADIUS proxy is far easier said than done but building out PI in this way would centralizes every FIDO2 interaction inside PrivacyIDEA’s service and be (IMO) simpler than the current “next to” model of simply augmenting your primary auth store, not to mention more future-proof.
